tools/libSwiftSyntaxParser/CMakeLists.txt

```diff
@@ -1,61 +1,112 @@
+    # -Wl,-x passes -x to the linker, which causes non-global symbols to not occur in the relocatable object file.
+    target_link_options("${target_name}" PRIVATE -Wl,-r -Wl,-x)
+    # Since we aren't producing a dylib anymore because of -Wl,-r, we should change the file extension
+    set_target_properties("${target_name}" PROPERTIES SUFFIX ".o")
+  endif()
+
+  if (SWIFT_LIBPARSER_VER)
+    set(SWIFTSYNTAX_PARSER_VERSION_STRING "${SWIFT_LIBPARSER_VER}")
   else()
-    set_target_properties(libSwiftSyntaxParser
-      PROPERTIES
-      VERSION ${SWIFTSYNTAX_PARSER_VERSION_STRING})
+    set(SWIFTSYNTAX_PARSER_VERSION_STRING "${SWIFT_COMPILER_VERSION}")
   endif()
-endif()
 
-set_property(TARGET libSwiftSyntaxParser APPEND_STRING PROPERTY
-  COMPILE_FLAGS " -fblocks")
-if(NOT CMAKE_SYSTEM_NAME STREQUAL Darwin)
-  target_link_libraries(libSwiftSyntaxParser PRIVATE
-    BlocksRuntime)
-endif()
+  if(SWIFTSYNTAX_PARSER_VERSION_STRING)
+    if("${CMAKE_SYSTEM_NAME}" STREQUAL "Darwin")
+      set(LIBPARSER_LINK_FLAGS " -Wl,-compatibility_version -Wl,1")
+      set(LIBPARSER_LINK_FLAGS "${LIBPARSER_LINK_FLAGS} -Wl,-current_version -Wl,${SWIFTSYNTAX_PARSER_VERSION_STRING}")
+      set_property(TARGET "${target_name}" APPEND_STRING PROPERTY
+                   LINK_FLAGS ${LIBPARSER_LINK_FLAGS})
+    else()
+      set_target_properties("${target_name}"
+        PROPERTIES
+        VERSION ${SWIFTSYNTAX_PARSER_VERSION_STRING})
+    endif()
+  endif()
+
+  set_property(TARGET "${target_name}" APPEND_STRING PROPERTY
+    COMPILE_FLAGS " -fblocks")
+  if(NOT CMAKE_SYSTEM_NAME STREQUAL Darwin)
+    target_link_libraries("${target_name}" PRIVATE
+      BlocksRuntime)
+  endif()
+endfunction()
+
+# Use an 'internal' name, this is primarily intended for SwiftSyntax to import.
+create_parser_library(libSwiftSyntaxParser _InternalSwiftSyntaxParser dylib)
 
 add_dependencies(parser-lib libSwiftSyntaxParser)
 swift_install_in_component(TARGETS libSwiftSyntaxParser
   ARCHIVE DESTINATION "lib${LLVM_LIBDIR_SUFFIX}/swift/${SWIFT_SDK_${SWIFT_HOST_VARIANT_SDK}_LIB_SUBDIR}" COMPONENT parser-lib
   LIBRARY DESTINATION "lib${LLVM_LIBDIR_SUFFIX}/swift/${SWIFT_SDK_${SWIFT_HOST_VARIANT_SDK}_LIB_SUBDIR}" COMPONENT parser-lib
   RUNTIME DESTINATION "bin" COMPONENT parser-lib)
 swift_install_in_component(DIRECTORY "${SWIFT_MAIN_INCLUDE_DIR}/swift-c/SyntaxParser/"
-                           DESTINATION "lib${LLVM_LIBDIR_SUFFIX}/swift/${SYNTAX_PARSER_LIB_NAME}"
+                           DESTINATION "lib${LLVM_LIBDIR_SUFFIX}/swift/_InternalSwiftSyntaxParser"
                            COMPONENT parser-lib)
+
+if("${CMAKE_SYSTEM_NAME}" STREQUAL "Darwin")
+  # We only need the static library to distribute it as an xcframework for SwiftSyntax.
+  # xcframework dependencies in SwiftPM are only supported on Darwin at the moment, so there's no need to build a static library for other platforms at the moment.
```

Comment:
This might be a bit misleading since other platforms could still benefit from this just not directly with swiftpm including it, but either way hopefully it's something we can do in the short term

Reply:
Reworded
```
  # TODO: Building the parser library as a static library has only been tested on Darwin so far because we only need the static library 
  # to distribute it as an xcframework for SwiftSyntax and xcframework dependencies in SwiftPM are only supported on Darwin at the moment.
  # Building and static linking the parser library on Linux might be possible, trivial.
```